File: Medium/147 - InsertionSortList/Solution.cs
// Question 147. Insertion Sort List (https://leetcode-cn.com/problems/insertion-sort-list/)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */

/// <summary>
/// 执行用时：140 ms, 在所有 C# 提交中击败了 46.67% 的用户
/// 内存消耗：25.4 MB, 在所有 C# 提交中击败了 13.33% 的用户
/// </summary>
public class ListNode 
{
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class Solution 
{
    public ListNode InsertionSortList(ListNode head) 
    {
        var sortedList = new ListNode(-1);
        var SORTED_LIST_HEAD = sortedList;

        while (head != null)
        {
            ListNode currentNode = head;
            head = head.next;

            sortedList = SORTED_LIST_HEAD;
            while (sortedList.next != null && sortedList.next.val < currentNode.val)
            {
                sortedList = sortedList.next;
            }
            currentNode.next = sortedList.next;
            sortedList.next = currentNode;
        }

        return SORTED_LIST_HEAD.next;
    }
}

File: Medium/147 - InsertionSortList/solution.ts
// Question 147. Insertion Sort List (https://leetcode-cn.com/problems/insertion-sort-list/)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * 执行用时：124 ms, 在所有 typescript 提交中击败了 44.44% 的用户
 * 内存消耗：41.9 MB, 在所有 typescript 提交中击败了 11.11% 的用户
 *
 * Acknowledgement: This solution is the TypeScript-version duplicate of the [official solution](https://leetcode-cn.com/problems/insertion-sort-list/solution/dui-lian-biao-jin-xing-cha-ru-pai-xu-by-leetcode-s/)
 *
 * @param {(ListNode | null)} head
 * @returns {(ListNode | null)}
 */
function insertionSortList(head: ListNode | null): ListNode | null {
  if (!head) {
    return head;
  }

  const dummyHead: ListNode = new ListNode(-1, head);
  let lastSorted: ListNode | null = head;
  let current: ListNode | null = head.next;

  while (current) {
    if (lastSorted && lastSorted.val <= current.val) {
      lastSorted = lastSorted.next;
    } else {
      let previous: ListNode | null = dummyHead;
      while (previous && previous.next && previous.next.val <= current.val) {
        previous = previous.next;
      }

      if (lastSorted && previous) {
        lastSorted.next = current.next;
        current.next = previous.next;
        previous.next = current;
      }
    }

    current = lastSorted?.next ?? null;
  }

  return dummyHead.next;
}
